Add an exported serve dashboard json

alanwguo opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
<img width="1904" alt="Screenshot 2023-03-16 at 7 57 01 PM" src="https://user-images.githubusercontent.com/711935/225801141-c6d5785a-f0e1-4576-9e3c-3e7082475d1e.png">
<img width="1904" alt="Screenshot 2023-03-16 at 7 57 05 PM" src="https://user-images.githubusercontent.com/711935/225801145-37e82150-4168-4f83-9061-a9883120ccab.png">


Refactor grafana_dashboard_factory to support multiple dashboards
Update serve http latency metric to use route instead route_prefix as tag

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented:
@alanwguo let's make the QPS graphs consistent. Right now there is total QPS then broken down by route but for errors it's broken down by deployment. I would suggest:

- For both, no need to have a separate plot for total vs. broken down, it can just be one plot w/ multiple lines including a total line (I think you can do that).
- Let's add by route and by deployment for both.

So I guess this means we should have four charts: QPS by route, QPS by deployment, error QPS by route, error QPS by deployment.

WDYT?

alanwguo commented:
> * For both, no need to have a separate plot for total vs. broken down, it can just be one plot w/ multiple lines including a total line (I think you can do that).
> * Let's add by route and by deployment for both.

I incorporated your feedback with the following changes:
1. got rid of the total qps and total latency charts and moved them to the "By route" charts.
2. Made the HTTP metrics and Deployment metrics more or less mirror each other.

Row 1 + 2: High level cluster utilization metric, request and error QPS and latencies by route
Row 3 + 4: High level deployments metric, request and error QPS and latencies by deployment
Row 5: Various useful hardware 


<img width="1904" alt="Screenshot 2023-03-17 at 4 08 10 PM" src="https://user-images.githubusercontent.com/711935/226067377-321d1bf0-8e4f-47bd-9244-0df4dc807264.png">
<img width="1904" alt="Screenshot 2023-03-17 at 4 08 13 PM" src="https://user-images.githubusercontent.com/711935/226067382-47ba03f6-4ec4-4656-a867-fe0b88808cbb.png">

edoakes reviewed (approved):
Looks awesome, thank you

scottsun94 commented:
The "queue size" in row 5 is also per deployment? Not sure if we make the title more explicit/consistent with others.

If we added those graphs into ray dashboard, it would be inside the application detail page?

rkooo567 commented on dashboard/modules/metrics/dashboards/default_dashboard_panels.py:
```diff
@@ -0,0 +1,366 @@
+# flake8: noqa E501
```
Is this file exactly the same as before?

alanwguo commented:
> I have two questions:
> 
> 1. The "queue size" in row 5 is also per deployment? Not sure if we make the title more explicit/consistent with others.
> 2. If we added those graphs into ray dashboard, it would be inside the application detail page?

1. Yes it is per deployment. I'll update the title

2. Right now we don't have metrics which are tagged by application. We won't be able to support application detail metrics until ray 2.5 I think. For now, we can stick with global level and deployment level.

scottsun94 commented:
> > I have two questions:
> > 
> > 1. The "queue size" in row 5 is also per deployment? Not sure if we make the title more explicit/consistent with others.
> > 2. If we added those graphs into ray dashboard, it would be inside the application detail page?
> 
> 1. Yes it is per deployment. I'll update the title
> 2. Right now we don't have metrics which are tagged by application. We won't be able to support application detail metrics until ray 2.5 I think. For now, we can stick with global level and deployment level.

SG. Thanks!

ericl reviewed (approved):
High level structure lgtm.

akshay-anyscale commented:
Are the queue size metrics only at the deployment level or do they include queues in HTTPProxy as well?

alanwguo commented:
> Are the queue size metrics only at the deployment level or do they include queues in HTTPProxy as well?

Only per deployment right now
